kernel/kernel-api/src/main/java/io/delta/kernel/internal/util/SchemaUtils.java

```diff
@@ -0,0 +1,125 @@
+/*
+ * Copyright (2023) The Delta Lake Project Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.delta.kernel.internal.util;
+
+import java.util.*;
+import java.util.stream.Collectors;
+import static java.lang.String.format;
+
+import io.delta.kernel.types.*;
+
+import static io.delta.kernel.internal.util.Preconditions.checkArgument;
+
+/**
+ * Utility methods for schema related operations such as validating the schema has no duplicate
+ * columns and the names contain only valid characters.
+ */
+public class SchemaUtils {
+
+    private SchemaUtils() {
+    }
+
+    /**
+     * Validate the schema. This method checks if the schema has no duplicate columns and the names
+     * contain only valid characters.
+     *
+     * @param schema                 the schema to validate
+     * @param isColumnMappingEnabled whether column mapping is enabled. When column mapping is
+     *                               enabled, the column names in the schema can contain special
+     *                               characters that are allowed as column names in the Parquet
+     *                               file
+     * @throws IllegalArgumentException if the schema is invalid
+     */
+    public static void validateSchema(StructType schema, boolean isColumnMappingEnabled) {
+        List<String> flattenColNames = flattenNestedFieldNames(schema);
+
+        // check there are no duplicate column names in the schema
+        Set<String> uniqueColNames = flattenColNames.stream()
+                .map(String::toLowerCase)
+                .collect(Collectors.toSet());
+
+        if (uniqueColNames.size() != flattenColNames.size()) {
+            String duplicateColumns = flattenColNames.stream()
+                    .map(String::toLowerCase)
+                    .filter(n -> !uniqueColNames.add(n))
```

Comment:
Isn't every element of `flattenColNames` already in `uniqueColNames` ?

so won't `uniqueColNames.add(n)` return false every time?

Reply:
You are right. This should be a new set. Added a test to match the error message exactly to contain only the duplicate column names.